File: public/packages/dotnet/caching/abstractions/LocalCacheOptions.cs
// -----------------------------------------------------------------------
// <copyright file="LocalCacheOptions.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Caching;

/// <summary>
/// Configuration knobs for an <see cref="ILocalCache"/> implementation.
/// Defaults are tuned for typical microservice workloads (100K entries,
/// 1-hour default TTL).
/// </summary>
/// <remarks>
/// Per-entry size accounting is intentionally NOT exposed — the default
/// implementation always counts entries as size 1 so <see cref="MaxEntries"/>
/// behaves as the literal entry-count cap most callers expect. Without
/// this discipline, <c>IMemoryCache</c>'s <c>SizeLimit</c> only triggers
/// eviction on system-wide memory pressure, making the configured ceiling
/// fictional.
/// </remarks>
public sealed class LocalCacheOptions
{
    /// <summary>
    /// Gets or sets the maximum number of entries before LRU-ish eviction
    /// kicks in. Default 100_000 — sized for production Edge / Files-class
    /// workloads where session-liveness, JWKS, token-exchange, and other
    /// per-process caches share a single <see cref="ILocalCache"/> singleton.
    /// At ~1 KB / entry that's ~100 MB process RSS worst-case. Per-service
    /// overrides remain possible via the <c>configure</c> delegate of
    /// <c>AddD2LocalCache</c>.
    /// </summary>
    public int MaxEntries { get; set; } = 100_000;

    /// <summary>
    /// Gets or sets the default TTL applied to entries written without an
    /// explicit expiration. Default 1 hour.
    /// </summary>
    public TimeSpan DefaultExpiration { get; set; } = TimeSpan.FromHours(1);

    /// <summary>
    /// Gets or sets the optional key prefix automatically prepended to
    /// every cache key. Useful when multiple caches share a process and
    /// the caller wants namespace isolation (e.g. <c>"jwks:"</c>) without
    /// coordinating keys explicitly. Default is empty (no prefix).
    /// </summary>
    public string KeyPrefix { get; set; } = string.Empty;
}

File: public/packages/typescript/caching/abstractions/src/local-cache-options.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

/**
 * Configuration knobs for an {@link ILocalCache} implementation.
 * Defaults are tuned for typical microservice workloads (100K entries,
 * 1-hour default TTL).
 *
 * Per-entry size accounting is intentionally NOT exposed — the default
 * implementation always counts entries as size 1 so `maxEntries`
 * behaves as the literal entry-count cap most callers expect.
 *
 * No options validation on this POCO (mirrors .NET — no range check /
 * ValidateOnStart). The factory does not fail on absurd values;
 * enforcement belongs in the local-default implementation.
 */
export interface LocalCacheOptions {
  /**
   * Maximum number of entries before LRU-ish eviction kicks in.
   * Default 100_000.
   */
  maxEntries: number;

  /**
   * Default TTL (milliseconds) applied to entries written without an
   * explicit expiration. Default 3_600_000 (1 hour).
   */
  defaultExpirationMs: number;

  /**
   * Optional key prefix automatically prepended to every cache key.
   * Useful when multiple caches share a process and the caller wants
   * namespace isolation (e.g. `"jwks:"`) without coordinating keys
   * explicitly. Default is empty string (no prefix).
   */
  keyPrefix: string;
}

// Numeric twin of .NET LocalCacheOptions (MaxEntries=100_000, DefaultExpiration=1h, KeyPrefix="").
export const LOCAL_CACHE_DEFAULTS: Readonly<LocalCacheOptions> = {
  maxEntries: 100_000,
  defaultExpirationMs: 3_600_000,
  keyPrefix: "",
};

/**
 * Builds a mutable {@link LocalCacheOptions} by merging an optional
 * partial over {@link LOCAL_CACHE_DEFAULTS}. Always returns a fresh
 * object (not shared with the defaults constant).
 */
export function createLocalCacheOptions(
  partial?: Partial<LocalCacheOptions>,
): LocalCacheOptions {
  return {
    maxEntries: partial?.maxEntries ?? LOCAL_CACHE_DEFAULTS.maxEntries,
    defaultExpirationMs:
      partial?.defaultExpirationMs ?? LOCAL_CACHE_DEFAULTS.defaultExpirationMs,
    keyPrefix: partial?.keyPrefix ?? LOCAL_CACHE_DEFAULTS.keyPrefix,
  };
}
